### DISCUSSION TITLE
Executing command on Cassandra test container

### DISCUSSION THREAD

**Discussion by YunusSarigul** [testcontainers/testcontainers-dotnet#436] [Q&A] [answered]
Hi, 
I created a Cassandra test container but could execute command on that container using container.execAsync function using following code. Is that anything wrong with what i am doing? I can run the same command on container via command line and it works fine btw. Any help would be great.

Code is basicly like following :

` Container = new TestcontainersBuilder<TestcontainersContainer>()
                .WithDockerEndpoint("unix:/var/run/docker.sock")
                .WithImage("cassandra:4.0.3")
                .WithName("sample-cassandra")
                .WithEnvironment("CASSANDRA_DC", "dc1")
                .WithExposedPort(9042)
                .WithExposedPort(9160)
                .WithPortBinding(9042, 9042)
                .WithPortBinding(9160, 9160)
                .WithWaitStrategy(Wait.ForUnixContainer().UntilPortIsAvailable(9042))
                .Build();

Container.StartAsync().ConfigureAwait(false).GetAwaiter().GetResult();
            
var createKeySpaceCommand =
                "cqlsh###CREATE KEYSPACE sample WITH replication = {'class':'SimpleStrategy', 'replication_factor' : 1};###";

Container.ExecAsync(createKeySpaceCommand.Split("###").ToList()).ConfigureAwait(false).GetAwaiter()
                .GetResult();`

**Comment by HofmeisterAn** [comment, marked as answer]
It's almost always the same. Usually, it's not enough to just wait for the port. The port is available before the container or the service, in that case Cassandra is ready. Use a custom wait strategy to determine when the service is running.

E. g. for PostgreSQL it looks like #412. [Here](https://github.com/HofmeisterAn/dotnet-testcontainers/issues/421#issuecomment-1055638676) is another example for azure-cosmos-emulator. For Cassandra, you could wait for a specific log message or something else.

Please add more information in the future to your question or discussion. What have you already tried? What error do you get? etc.

**Reply by YunusSarigul** [reply]
Thanks for the answer. Changing wait strategy did not actually solved my problem by itself. I think problem was about the command I was trying to execute.I give up trying to execute command on container and used CassandraCSharpDriver and that works fine.

**Reply by HofmeisterAn** [reply]
At least, you need to escape your `cqlsh` statement, like:

```csharp
var execResult = await this.testcontainer.ExecAsync(new[] { "cqlsh", "\"CREATE KEYSPACE sample WITH replication = {'class':'SimpleStrategy', 'replication_factor' : 1};\"" })
  .ConfigureAwait(false);
```

Then check the `execResult`. If it still does not work, it contains the exit code and error message.

**Reply by YunusSarigul** [reply]
Worked with escape characters ! Thank you.
